pixeltable/catalog/catalog.py

```diff
@@ -114,17 +120,96 @@ def clear(cls) -> None:
+                f'{effective_version} != {tbl_version.effective_version} for id {tbl_id}'
+            )
+            assert len(tbl_version.mutable_views) == 0 or tbl_version.is_mutable, (
+                f'snapshot_id={tbl_version.id} mutable_views={tbl_version.mutable_views}'
+            )
+
+            if tbl_version.is_view and tbl_version.is_mutable:
+                # make sure this mutable view is recorded in a mutable base
+                base = tbl_version.base
+                assert base is not None
+                if base.effective_version is None:
+                    assert (base.id, None) in self._tbl_versions
+                    assert TableVersionHandle.create(tbl_version) in self._tbl_versions[base.id, None].mutable_views
+
+            if len(tbl_version.mutable_views) > 0:
+                # make sure we also loaded mutable view metadata, which is needed to detect column dependencies
+                for v in tbl_version.mutable_views:
+                    assert v.effective_version is None, f'{v.id}:{v.effective_version}'
+
+    @contextmanager
+    def begin_xact(self, *, tbl_id: Optional[UUID] = None, for_write: bool = False) -> Iterator[sql.Connection]:
+        """
+        Return a context manager that yields a connection to the database. Idempotent.
+
+        It is mandatory to call this method, not Env.begin_xact(), if the transaction accesses any table data
+        or metadata.
+        """
+        if not Env.get().in_xact:
+            num_retries = 0
+            while True:
+                try:
+                    with Env.get().begin_xact() as conn:
+                        if tbl_id is not None and for_write:
+                            # X-lock Table record
+                            conn.execute(
+                                sql.select(schema.Table).where(schema.Table.id == tbl_id).with_for_update(nowait=True)
+                            )
+                            conn.execute(sql.update(schema.Table).values(lock_dummy=1).where(schema.Table.id == tbl_id))
+
+                        self._in_write_xact = for_write
+                        yield conn
+                        return
+                except sql.exc.DBAPIError as e:
+                    if isinstance(e.orig, (psycopg.errors.SerializationFailure, psycopg.errors.LockNotAvailable)) and (
+                        num_retries < _MAX_RETRIES or _MAX_RETRIES == 0
+                    ):
+                        num_retries += 1
+                        _logger.debug(f'Retrying ({num_retries}) after {type(e.orig)}')
+                        time.sleep(random.uniform(0.1, 0.5))
+                    else:
+                        raise
+                finally:
+                    self._in_write_xact = False
+
+                    # invalidate cached current TableVersion instances
+                    for tv in self._tbl_versions.values():
+                        tv.is_validated = False
```

Comment:
I'd be more comfortable with a `tv.invalidate()` method (instead of reaching inside the class to mess with its member variables)